core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProviderBuilder.scala

```diff
@@ -39,8 +40,24 @@ class DataRepoEntityProviderBuilder(workspaceManagerDAO: WorkspaceManagerDAO, da
       snapshotModel <- Try(dataRepoDAO.getSnapshot(snapshotId, requestArguments.userInfo.accessToken)).recoverWith {
         case notFound: DatarepoApiException if notFound.getCode == StatusCodes.NotFound.intValue =>
           Failure(new DataEntityException(s"Snapshot id $snapshotId does not exist or you do not have access", notFound))
-        case forbidden: DatarepoApiException if forbidden.getCode == StatusCodes.Forbidden.intValue =>
-          Failure(new DataEntityException(s"Snapshot id $snapshotId exists but access was denied", forbidden))
+
+        // Data Repo returns 401 Unauthorized if the user has a valid token but does not have the proper permission on the snapshot
+        case forbidden: DatarepoApiException if forbidden.getCode == StatusCodes.Forbidden.intValue || forbidden.getCode == StatusCodes.Unauthorized.intValue =>
+          // attempt to extract buried message
+          // raw response looks like: {"message":"User 'davidan.dev2@gmail.com' does not have required action: read_data","errorDetail":[]}
+          // TODO: probably want this logic elsewhere, extract to a reusable method?
+          import spray.json._
+          val innerErrorMessage: String = Try(forbidden.getMessage.parseJson.asJsObject.fields.get("message"))
+            .toOption.flatten.collect {
+              case jss:JsString => jss.value
+            }
+            .getOrElse("[error could not be deserialized]")
```

Comment:
can we also throw the raw response in a logger.warn message if we hit this case? i am just trying to avoid any potential head-scratching scenarios in the future when debugging

Reply:
see the `logger.warn` at line 59 ... is that what you're looking for? I explicitly didn't want to throw the entire response to the end user, because it can be a long ugly stack trace, but I did log it.

Reply:
ah I misread the items being logged in line 59. yes it's already doing what i was asking for!